### ISSUE TITLE
LV2 Oscilloscope background grids do not get rendered when LC_NUMERIC is set to a locale, that does not use dot as decimal point

### ISSUE THREAD

**Issue by morten71** [sadko4u/lsp-plugins#352]
On linux, the background grids on the oscilloscope graphs do not get rendered when running as LV2 plugins, but they do get rendered correctly when running the stand-alone jack versions.

I have tried my best to track down the problem, and this is what I have been able to come up with:

The problem seems to come from the float values that get calculated for the v attribute in the marker tags, in the ui xml files. For example in lsp-plugins-oscilloscope/res/main/ui/util/oscilloscope/x1.xml:

```
<ui:for id="i" first="-31" last="31">
    <ui:if test=":i mod 4">
        <marker v="${:i * 0.03125}" ox="1" oy="0" color="graph_alt_4"/>
        <marker v="${:i * 0.03125}" color="graph_alt_4"/>
    </ui:if>
</ui:for>
```

After they have been calculated, these float values get converted to an LSPString before they end up getting set on the Marker object's sValue field. The conversion happens  on line 2540 in lsp-runtime-lib/src/main/runtime/LSPString.cpp in a call to vasprintf.

I tried inserting a call at the start of the vfmt_ascii function, like this (after having included clocale of course):

`char* lres = std::setlocale(LC_NUMERIC, NULL);`

When I run the code as an LV2 plugin, this call returns my system's LC_NUMERIC locale "da_DK.utf8", and when I run it as the stand-alone jack version, the call returns "C".

The result is that the float values get set on the Marker object as strings with comma as the decimal point, causing the values to fail being parsed, which then results in v being set to zero for all the marker lines. I guess maybe they do get drawn, all in the center, overlapping the axes :D

I stopped tracking the problem after this, so I have no idea why the locale gets set to C (like it's intended to be, I think) for the conversion in the jack plugins, but not in the LV2 plugins.

Now I just set LC_NUMERIC to C when I run my plugin hosts, but it would be nice if I didn't have to. Also for a long time I thought it was an intentional design for the oscilloscopes, until I decided that that was too weird, and started to investigate. I think it's a shame if other people run into the same situation, and just think that the oscilloscopes are meant to look barren and weird like that.

**Comment by sadko4u** [comment]
Hello!
Interesting case.
Thank you for detailed investigation. I was able to reproduce the problem.

I've committed changes to all places where floating point values become formatted and now build from `devel` branch should work fine for you. Please verify if it is possible.

**Comment by morten71** [comment]
Yes, I can confirm that it does work now! Thanks for the quick fix :-)

**Comment by morten71** [comment]
There is one little detail I have noticed, that is probably very related: All the UIs present float values with decimal points respecting the system's locale, which in my case is with commas. But when double clicking on controllers to input values manually (which I always do, because I'm a freak that likes precise values, even though it makes no difference), the input field only accepts dot as the decimal point.

This is absolutely not a problem for me, and I have never considered it a bug. But it is maybe a slightly weird inconsistency.

Please ignore this, if you don't think it's important. I don't think it's very important myself, but now that this float conversion problem came up, I thought I would mention it.

**Comment by sadko4u** [comment]
That's because we use our own number parsers which respect only dot as a decimal symbol. These parsers are the part of expression language which you already have seen in XML files. At this moment I've updated the parsing code. Maybe there is still somewhere the formatting code which is using the system default locale.

**Comment by morten71** [comment]
That makes perfect sense. Thanks for replying!

I think I was just mostly curious about the difference in behavior between the LV2 and jack versions, but I can figure out why that happens if I want to. But I don't think it's important enough, so I think I will just get back to using the plugins 😅

**Comment by sadko4u** [comment]
Fixed in 1.2.14!
